File: ios-native/PearBrowser/Sources/UI/Screens/MySitesScreen.swift
//  PearBrowser — MySitesScreen.swift
//
//  SwiftUI mirror of app/screens/MySitesScreen.tsx. Lists the user's
//  personal websites (published as Hyperdrives), lets them create a
//  new one, edit it, publish it, share the hyper:// URL, or delete.
//
//  Uses PearRPC's CMD_LIST_SITES / CMD_CREATE_SITE / CMD_PUBLISH_SITE /
//  CMD_DELETE_SITE.

import SwiftUI

struct SiteRecord: Identifiable, Hashable {
    let id: String
    let keyHex: String
    let name: String
    let published: Bool
    let url: String
}

struct MySitesScreen: View {
    let onEdit: (String) -> Void
    let onPreview: (String) -> Void
    let onCreateNew: (String) -> Void
    let onBack: () -> Void

    @Environment(\.pearRPC) private var rpc
    @State private var sites: [SiteRecord] = []
    @State private var newName: String = ""
    @State private var creating = false
    @State private var loading = true
    @State private var errorMessage: String? = nil
    @State private var publishResult: String? = nil
    @State private var shareItem: String? = nil
    @State private var siteToDelete: SiteRecord? = nil

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader("My Sites", onBack: onBack)

            ScrollView {
                VStack(alignment: .leading, spacing: 14) {
                    Text("Create and publish P2P websites")
                        .font(.system(size: 13))
                        .foregroundStyle(PearColors.textSecondary)

                    createRow

                    if let errorMessage {
                        errorBox(errorMessage) { Task { await load() } }
                    }

                    if loading {
                        ProgressView().tint(PearColors.accent).padding(.top, 20)
                    } else if sites.isEmpty {
                        emptyState
                    } else {
                        ForEach(sites) { site in
                            siteCard(site)
                        }
                    }
                }
                .padding(16)
            }
        }
        .background(PearColors.bg)
        .task { await load() }
        .alert("Site published!", isPresented: Binding(get: { publishResult != nil }, set: { if !$0 { publishResult = nil } })) {
            Button("OK", role: .cancel) {}
        } message: {
            if let publishResult { Text("Your site is live at:\n\nhyper://\(publishResult.prefix(16))…") }
        }
        .alert("Delete site?", isPresented: Binding(get: { siteToDelete != nil }, set: { if !$0 { siteToDelete = nil } })) {
            Button("Cancel", role: .cancel) {}
            Button("Delete", role: .destructive) {
                if let site = siteToDelete { delete(site) }
            }
        } message: {
            Text("\"\(siteToDelete?.name ?? "")\" will be removed from this device.")
        }
        .sheet(isPresented: Binding(get: { shareItem != nil }, set: { if !$0 { shareItem = nil } })) {
            if let item = shareItem {
                ShareSheet(activityItems: [item])
            }
        }
    }

    private var createRow: some View {
        HStack(spacing: 8) {
            TextField("Site name…", text: $newName)
                .font(.system(size: 16))
                .foregroundStyle(PearColors.textPrimary)
                .padding(12)
                .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 12))
            Button {
                if let onCreateNew = onCreateNewIfValid() {
                    onCreateNew()
                }
            } label: {
                if creating {
                    ProgressView().tint(PearColors.bg)
                } else {
                    Text("Create")
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundStyle(PearColors.bg)
                }
            }
            .padding(.horizontal, 20)
            .padding(.vertical, 12)
            .background(creating || newName.trimmingCharacters(in: .whitespaces).isEmpty
                        ? PearColors.surface
                        : PearColors.accent,
                        in: RoundedRectangle(cornerRadius: 12))
            .disabled(creating || newName.trimmingCharacters(in: .whitespaces).isEmpty)
        }
    }

    private func onCreateNewIfValid() -> (() -> Void)? {
        let trimmed = newName.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return nil }
        return {
            newName = ""
            onCreateNew(trimmed)
        }
    }

    private func errorBox(_ message: String, retry: @escaping () -> Void) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Could not load sites")
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(PearColors.error)
            Text(message)
                .font(.system(size: 12))
                .foregroundStyle(PearColors.textSecondary)
            Button("Retry", action: retry)
                .font(.system(size: 13, weight: .medium))
                .foregroundStyle(PearColors.accent)
        }
        .padding(14)
        .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 12))
    }

    private var emptyState: some View {
        VStack(spacing: 12) {
            Text("</>").font(.system(size: 36, design: .monospaced)).foregroundStyle(PearColors.accent)
            Text("No sites yet")
                .font(.system(size: 18, weight: .semibold))
                .foregroundStyle(PearColors.textPrimary)
            Text("Create your first P2P website above. It will be served from your phone and available to anyone on the network.")
                .font(.system(size: 13))
                .foregroundStyle(PearColors.textSecondary)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 40)
    }

    private func siteCard(_ site: SiteRecord) -> some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack(spacing: 12) {
                ZStack {
                    RoundedRectangle(cornerRadius: 10).fill(PearColors.surfaceElevated).frame(width: 40, height: 40)
                    Text(site.name.first.map { String($0).uppercased() } ?? "?")
                        .font(.system(size: 18, weight: .bold))
                        .foregroundStyle(PearColors.accent)
                }
                VStack(alignment: .leading, spacing: 2) {
                    Text(site.name)
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundStyle(PearColors.textPrimary)
                    Text("\(site.published ? "Live" : "Draft") · hyper://\(site.keyHex.prefix(8))…")
                        .font(.system(size: 11, design: .monospaced))
                        .foregroundStyle(PearColors.textMuted)
                }
                Spacer()
                Text(site.published ? "Live" : "Draft")
                    .font(.system(size: 11, weight: .semibold))
                    .foregroundStyle(site.published ? PearColors.success : PearColors.textMuted)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 3)
                    .background(site.published ? Color(red: 0.09, green: 0.39, blue: 0.27) : PearColors.surfaceElevated,
                                in: RoundedRectangle(cornerRadius: 8))
            }
            HStack(spacing: 8) {
                actionButton("Edit") { onEdit(site.id) }
                actionButton("Preview") { onPreview(site.url) }
                if site.published {
                    actionButton("Share", prominent: false) { shareItem = site.url }
                } else {
                    actionButton("Publish", prominent: true) { publish(site.id) }
                }
                actionButton("Delete", destructive: true) { siteToDelete = site }
            }
        }
        .padding(16)
        .background(PearColors.surface, in: RoundedRectangle(cornerRadius: 12))
    }

    private func actionButton(_ title: String, prominent: Bool = false, destructive: Bool = false, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(title)
                .font(.system(size: 13, weight: .medium))
                .foregroundStyle(destructive ? PearColors.error
                                  : (prominent ? PearColors.bg : PearColors.textSecondary))
                .padding(.horizontal, 12)
                .padding(.vertical, 6)
                .background(prominent ? PearColors.accent : PearColors.surfaceElevated,
                            in: RoundedRectangle(cornerRadius: 8))
        }
        .buttonStyle(.plain)
    }

    // MARK: - Actions

    private func load() async {
        loading = true
        errorMessage = nil
        defer { loading = false }
        guard let rpc else { sites = []; return }
        do {
            let resp = try await rpc.request(Cmd.LIST_SITES)
            if let arr = resp as? [[String: Any]] {
                sites = arr.compactMap { toSiteRecord($0) }
            } else if let obj = resp as? [String: Any], let arr = obj["sites"] as? [[String: Any]] {
                sites = arr.compactMap { toSiteRecord($0) }
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func toSiteRecord(_ dict: [String: Any]) -> SiteRecord? {
        guard let id = dict["siteId"] as? String ?? dict["id"] as? String,
              let keyHex = dict["keyHex"] as? String,
              let name = dict["name"] as? String else { return nil }
        let published = (dict["published"] as? Bool) ?? false
        let url = (dict["url"] as? String) ?? "hyper://\(keyHex)"
        return SiteRecord(id: id, keyHex: keyHex, name: name, published: published, url: url)
    }

    private func publish(_ siteId: String) {
        guard let rpc else { return }
        Task {
            do {
                let resp = try await rpc.request(Cmd.PUBLISH_SITE, data: ["siteId": siteId])
                if let obj = resp as? [String: Any], let key = obj["keyHex"] as? String {
                    publishResult = key
                }
                await load()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    private func delete(_ site: SiteRecord) {
        guard let rpc else { return }
        Task {
            do {
                _ = try await rpc.request(Cmd.DELETE_SITE, data: ["siteId": site.id])
                await load()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }
}

// MARK: - ShareSheet wrapper

struct ShareSheet: UIViewControllerRepresentable {
    let activityItems: [Any]
    func makeUIViewController(context: Context) -> UIActivityViewController {
        UIActivityViewController(activityItems: activityItems, applicationActivities: nil)
    }
    func updateUIViewController(_ uiViewController: UIActivityViewController, context: Context) {}
}

File: android-native/app/src/main/java/com/pearbrowser/app/ui/screens/MySitesScreen.kt
package com.pearbrowser.app.ui.screens

import android.content.Intent
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.pearbrowser.app.rpc.LocalPearRpc
import com.pearbrowser.app.rpc.PearSite
import com.pearbrowser.app.ui.theme.PearColors
import kotlinx.coroutines.launch
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive

/**
 * MySitesScreen — mirror of ios-native `MySitesScreen.swift`.
 *
 * Lists the user's sites (CMD_LIST_SITES), creates new ones (routes into the
 * template picker, which issues CMD_CREATE_SITE), and per site: edit,
 * preview, publish/unpublish toggle (CMD_PUBLISH_SITE / CMD_UNPUBLISH_SITE),
 * share the hyper:// URL, and delete (CMD_DELETE_SITE).
 */
@Composable
fun MySitesScreen(
    onEdit: (PearSite) -> Unit,
    onPreview: (String) -> Unit,
    onCreateNew: (String) -> Unit,
    onBack: () -> Unit,
) {
    val context = LocalContext.current
    val rpc = LocalPearRpc.current
    val scope = rememberCoroutineScope()

    var sites by remember { mutableStateOf<List<PearSite>>(emptyList()) }
    var loading by remember { mutableStateOf(true) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var notice by remember { mutableStateOf<String?>(null) }
    var newName by remember { mutableStateOf("") }
    var busySiteId by remember { mutableStateOf<String?>(null) }
    var siteToDelete by remember { mutableStateOf<PearSite?>(null) }

    suspend fun load() {
        loading = true
        errorMessage = null
        val client = rpc
        if (client == null) {
            sites = emptyList()
            errorMessage = "P2P engine is not connected yet"
            loading = false
            return
        }
        try {
            sites = client.listSites()
        } catch (e: Throwable) {
            errorMessage = e.message ?: "Could not load sites"
            sites = emptyList()
        } finally {
            loading = false
        }
    }

    fun refresh() {
        scope.launch { load() }
    }

    fun publish(site: PearSite) {
        val client = rpc ?: return
        busySiteId = site.siteId
        errorMessage = null
        notice = null
        scope.launch {
            try {
                val resp = client.publishSite(site.siteId)
                val key = resp["keyHex"]?.jsonPrimitive?.contentOrNull ?: site.keyHex
                notice = "Site published! Live at hyper://${key.take(16)}…"
                load()
            } catch (e: Throwable) {
                errorMessage = e.message ?: "Could not publish site"
            } finally {
                busySiteId = null
            }
        }
    }

    fun unpublish(site: PearSite) {
        val client = rpc ?: return
        busySiteId = site.siteId
        errorMessage = null
        notice = null
        scope.launch {
            try {
                client.unpublishSite(site.siteId)
                notice = "Site unpublished. It stays on this device as a draft."
                load()
            } catch (e: Throwable) {
                errorMessage = e.message ?: "Could not unpublish site"
            } finally {
                busySiteId = null
            }
        }
    }

    fun delete(site: PearSite) {
        val client = rpc ?: return
        busySiteId = site.siteId
        errorMessage = null
        notice = null
        scope.launch {
            try {
                client.deleteSite(site.siteId)
                load()
            } catch (e: Throwable) {
                errorMessage = e.message ?: "Could not delete site"
            } finally {
                busySiteId = null
            }
        }
    }

    fun share(site: PearSite) {
        try {
            val shareIntent = Intent(Intent.ACTION_SEND).apply {
                type = "text/plain"
                putExtra(Intent.EXTRA_TEXT, site.url)
            }
            context.startActivity(Intent.createChooser(shareIntent, "Share site"))
        } catch (e: Throwable) {
            errorMessage = e.message ?: "Could not share site"
        }
    }

    LaunchedEffect(rpc) { load() }

    Column(
        Modifier
            .fillMaxSize()
            .background(PearColors.Bg),
    ) {
        Row(
            Modifier
                .fillMaxWidth()
                .background(PearColors.Bg)
                .padding(horizontal = 10.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            TextButton(onClick = onBack) {
                Text("< Back", color = PearColors.Accent, fontSize = 14.sp)
            }
            Text(
                "My Sites",
                color = PearColors.TextPrimary,
                fontSize = 22.sp,
                fontWeight = FontWeight.Bold,
            )
        }

        Column(
            Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(16.dp),
        ) {
            Text(
                "Create and publish P2P websites",
                color = PearColors.TextSecondary,
                fontSize = 13.sp,
            )
            Spacer(Modifier.height(12.dp))

            // Create row — routes into the template picker (iOS onCreateNew).
            Row(verticalAlignment = Alignment.CenterVertically) {
                TextField(
                    value = newName,
                    onValueChange = { newName = it },
                    placeholder = { Text("Site name…", color = PearColors.TextMuted, fontSize = 16.sp) },
                    singleLine = true,
                    textStyle = TextStyle(color = PearColors.TextPrimary, fontSize = 16.sp),
                    colors = TextFieldDefaults.colors(
                        focusedContainerColor = PearColors.Surface,
                        unfocusedContainerColor = PearColors.Surface,
                        focusedIndicatorColor = Color.Transparent,
                        unfocusedIndicatorColor = Color.Transparent,
                        cursorColor = PearColors.Accent,
                        focusedTextColor = PearColors.TextPrimary,
                        unfocusedTextColor = PearColors.TextPrimary,
                    ),
                    shape = RoundedCornerShape(12.dp),
                    modifier = Modifier.weight(1f),
                )
                Spacer(Modifier.size(8.dp))
                val canCreate = newName.trim().isNotEmpty()
                Text(
                    "Create",
                    color = if (canCreate) PearColors.Bg else PearColors.TextMuted,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.SemiBold,
                    modifier = Modifier
                        .background(
                            if (canCreate) PearColors.Accent else PearColors.Surface,
                            RoundedCornerShape(12.dp),
                        )
                        .clickable(enabled = canCreate) {
                            val name = newName.trim()
                            newName = ""
                            onCreateNew(name)
                        }
                        .padding(horizontal = 20.dp, vertical = 12.dp),
                )
            }

            errorMessage?.let {
                Spacer(Modifier.height(12.dp))
                Column(
                    Modifier
                        .fillMaxWidth()
                        .background(PearColors.Surface, RoundedCornerShape(12.dp))
                        .padding(14.dp),
                ) {
                    Text(
                        "Could not load sites",
                        color = PearColors.Error,
                        fontSize = 13.sp,
                        fontWeight = FontWeight.SemiBold,
                    )
                    Text(it, color = PearColors.TextSecondary, fontSize = 12.sp)
                    Text(
                        "Retry",
                        color = PearColors.Accent,
                        fontSize = 13.sp,
                        fontWeight = FontWeight.Medium,
                        modifier = Modifier
                            .clickable { refresh() }
                            .padding(vertical = 4.dp),
                    )
                }
            }

            notice?.let {
                Spacer(Modifier.height(12.dp))
                Text(it, color = PearColors.Success, fontSize = 12.sp)
            }

            Spacer(Modifier.height(12.dp))
            when {
                loading -> Row(
                    Modifier.fillMaxWidth().padding(top = 20.dp),
                    horizontalArrangement = Arrangement.Center,
                ) {
                    CircularProgressIndicator(color = PearColors.Accent)
                }
                sites.isEmpty() && errorMessage == null -> Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    modifier = Modifier.fillMaxWidth().padding(vertical = 40.dp),
                ) {
                    Text(
                        "</>",
                        color = PearColors.Accent,
                        fontSize = 36.sp,
                        fontFamily = FontFamily.Monospace,
                    )
                    Spacer(Modifier.height(12.dp))
                    Text(
                        "No sites yet",
                        color = PearColors.TextPrimary,
                        fontSize = 18.sp,
                        fontWeight = FontWeight.SemiBold,
                    )
                    Spacer(Modifier.height(8.dp))
                    Text(
                        "Create your first P2P website above. It will be served from your phone and available to anyone on the network.",
                        color = PearColors.TextSecondary,
                        fontSize = 13.sp,
                        textAlign = TextAlign.Center,
                    )
                }
                else -> sites.forEach { site ->
                    SiteCard(
                        site = site,
                        busy = busySiteId == site.siteId,
                        onEdit = { onEdit(site) },
                        onPreview = { onPreview(site.url) },
                        onPublish = { publish(site) },
                        onUnpublish = { unpublish(site) },
                        onShare = { share(site) },
                        onDelete = { siteToDelete = site },
                    )
                    Spacer(Modifier.height(12.dp))
                }
            }
        }
    }

    siteToDelete?.let { site ->
        AlertDialog(
            onDismissRequest = { siteToDelete = null },
            containerColor = PearColors.Surface,
            title = { Text("Delete site?", color = PearColors.TextPrimary, fontWeight = FontWeight.Bold) },
            text = {
                Text(
                    "\"${site.name}\" will be removed from this device.",
                    color = PearColors.TextSecondary,
                    fontSize = 14.sp,
                )
            },
            confirmButton = {
                TextButton(onClick = {
                    siteToDelete = null
                    delete(site)
                }) {
                    Text("Delete", color = PearColors.Error, fontWeight = FontWeight.Bold)
                }
            },
            dismissButton = {
                TextButton(onClick = { siteToDelete = null }) {
                    Text("Cancel", color = PearColors.TextSecondary)
                }
            },
        )
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun SiteCard(
    site: PearSite,
    busy: Boolean,
    onEdit: () -> Unit,
    onPreview: () -> Unit,
    onPublish: () -> Unit,
    onUnpublish: () -> Unit,
    onShare: () -> Unit,
    onDelete: () -> Unit,
) {
    Column(
        Modifier
            .fillMaxWidth()
            .background(PearColors.Surface, RoundedCornerShape(12.dp))
            .padding(16.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center,
                modifier = Modifier
                    .size(40.dp)
                    .background(PearColors.SurfaceElevated, RoundedCornerShape(10.dp)),
            ) {
                Text(
                    site.name.firstOrNull()?.uppercase() ?: "?",
                    color = PearColors.Accent,
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                )
            }
            Spacer(Modifier.size(12.dp))
            Column(Modifier.weight(1f)) {
                Text(
                    site.name,
                    color = PearColors.TextPrimary,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.SemiBold,
                )
                Text(
                    "${if (site.published) "Live" else "Draft"} · hyper://${site.keyHex.take(8)}…",
                    color = PearColors.TextMuted,
                    fontSize = 11.sp,
                    fontFamily = FontFamily.Monospace,
                )
            }
            Text(
                if (site.published) "Live" else "Draft",
                color = if (site.published) PearColors.Success else PearColors.TextMuted,
                fontSize = 11.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier
                    .background(
                        if (site.published) Color(0xFF176345) else PearColors.SurfaceElevated,
                        RoundedCornerShape(8.dp),
                    )
                    .padding(horizontal = 8.dp, vertical = 3.dp),
            )
        }
        Spacer(Modifier.height(12.dp))
        FlowRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            SiteAction("Edit", onClick = onEdit)
            SiteAction("Preview", onClick = onPreview)
            if (site.published) {
                SiteAction(if (busy) "…" else "Unpublish", enabled = !busy, onClick = onUnpublish)
                SiteAction("Share", onClick = onShare)
            } else {
                SiteAction(if (busy) "…" else "Publish", prominent = true, enabled = !busy, onClick = onPublish)
            }
            SiteAction("Delete", destructive = true, enabled = !busy, onClick = onDelete)
        }
    }
}

@Composable
private fun SiteAction(
    label: String,
    prominent: Boolean = false,
    destructive: Boolean = false,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    val color = when {
        !enabled -> PearColors.TextMuted
        destructive -> PearColors.Error
        prominent -> PearColors.Bg
        else -> PearColors.TextSecondary
    }
    Text(
        label,
        color = color,
        fontSize = 13.sp,
        fontWeight = FontWeight.Medium,
        modifier = Modifier
            .background(
                if (prominent) PearColors.Accent else PearColors.SurfaceElevated,
                RoundedCornerShape(8.dp),
            )
            .clickable(enabled = enabled, onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 6.dp),
    )
}
